### ISSUE TITLE
support `dtype` kwarg for grouped reductions

### ISSUE THREAD

**Issue by dcherian** [numbagg/numbagg#121]

**Comment by max-sixty** [comment]
Is `.sum(dtype=foo)` equivalent to calling `.astype(foo)` on the input? On the result?

**Comment by dcherian** [comment]
No, `dtype` is the `dtype` of the accumulator, so for example, you might want to use `dtype=np.float64` to more accurately reduce `np.float32` values.

Here's a PR with useful tests and discussion: https://github.com/xarray-contrib/flox/pull/131

**Comment by max-sixty** [comment]
I see. I think this is possible, though not trivial — currently we have a [list of candidate signatures](https://github.com/numbagg/numbagg/blob/19de7e8664bd72c5cb707b03a31bf40e1ffa9865/numbagg/grouped.py#L13) which numba will JIT compile for. We could generate new ones at runtime, but I'm not sure how numba handles a many-to-one mapping of inputs-to-outputs; so possibly we'd need to compile a new function...

**Comment by dcherian** [comment]
Ah I think this is lower priority. We can easily raise if dtype is provided and ask them to use another "engine"

**Comment by max-sixty** [comment]
> We can easily raise if dtype is provided and ask them to use another "engine"

Yes or to convert first (albeit requires a copy)

**Comment by max-sixty** [comment]
One question out of https://github.com/numbagg/numbagg/pull/128:

To what extent do we want to convert types? Currently we always have the same input type as output type... ...except for `bool` which we convert to ints (ref #128 )

But often that's not what we want — e.g. it is useful to have a float mean of ints, rather than an "int mean" (sum and then integer division).

A few choices:
- Always keep types — current state (except for bools...)
- Always convert to float64 — pandas' approach IIUC
- Allow passing a `dtype` arg — great but requires some work

Unless folks disagree, I think we'll stay at the current state and see if we get traction and folks ask for a change...

**Comment by dcherian** [comment]
> To what extent do we want to convert types?

I think we can just ask the user to do so...

> I think we'll stay at the current state and see if we get traction and folks ask for a change

IMO this is OK.

> A few choices:
>    Always keep types — current state (except for bools...)
>   Always convert to float64 — pandas' approach IIUC
>    Allow passing a dtype arg — great but requires some work

Perhaps @shoyer has an opinion
